src/kessel/rbac/v2/__init__.py

```diff
@@ -0,0 +1,151 @@
+        domain: The domain or organization the user belongs to
+    
+    Returns:
+        ResourceReference: A ResourceReference pointing to the user's principal resource
+    """
+    return ResourceReference(
+        resource_type="principal",
+        resource_id=f"{domain}/{id}",
+        reporter=ReporterReference(type="rbac")
+    )
+
+
+def role_resource(resource_id: str) -> ResourceReference:
+    """
+    Function to create a ResourceReference for a role.
+    Returns a protobuf ResourceReference configured for RBAC role resources.
+    
+    Args:
+        resource_id: The role identifier
+    
+    Returns:
+        ResourceReference: A ResourceReference for the role
+    """
+    return ResourceReference(
+        resource_type="role",
+        resource_id=resource_id,
+        reporter=ReporterReference(type="rbac")
+    )
+
+
+def workspace_resource(resource_id: str) -> ResourceReference:
+    """
+    Function to create a ResourceReference for a workspace.
+    Returns a protobuf ResourceReference configured for RBAC workspace resources.
+    
+    Args:
+        resource_id: The workspace identifier
+    
+    Returns:
+        ResourceReference: A ResourceReference for the workspace
+    """
+    return ResourceReference(
+        resource_type="workspace",
+        resource_id=resource_id,
+        reporter=ReporterReference(type="rbac")
+    )
+
+
+def principal_subject(id: str, domain: str) -> SubjectReference:
+    """
+    Creates a SubjectReference for a user principal based on user ID and domain.
+    This is a convenience function that wraps principal_resource to create a subject reference.
+    
+    Args:
+        id: The user identifier
+        domain: The domain or organization the user belongs to
+    
+    Returns:
+        SubjectReference: A SubjectReference for the user's principal
+    """
+    return SubjectReference(
+        resource=principal_resource(id, domain)
+    )
+
+
+def subject(resource_ref: ResourceReference, relation: Optional[str] = None) -> SubjectReference:
+    """
+    Creates a SubjectReference from a ResourceReference and an optional relation.
+    This function allows you to easily create a subject reference.
+    
+    Args:
+        resource_ref: The resource reference that identifies the subject
+        relation: Optional relation that points to a set of subjects (e.g., "members", "owners")
+    
+    Returns:
+        SubjectReference: A reference to a Subject or, if a relation is provided, a Subject Set.
+    """
+    if relation is not None:
+        return SubjectReference(
+            resource=resource_ref,
+            relation=relation
+        )
+    else:
+        return SubjectReference(
+            resource=resource_ref
+        )
+
+
+def list_workspaces(
+    subject: SubjectReference, 
+    relation: str, 
+    inventory: KesselInventoryServiceStub
+) -> AsyncIterator[StreamedListObjectsResponse] | Iterable[StreamedListObjectsResponse]:
```

Comment:
I think we should build the request without calling the client itself - it makes things more messy with the return types, async and sync stuff.

Reply:
Well - now I see that this API is paginated, and I wonder if this function could handle the pagination (i.e. get all resources by going over all the pages).

As per my concern above, what about creating an async version of this, like the `ClientBuilder.build` method. i.e.  `list_workspaces_async`.

WDYT?

Reply:
Yeah a separate `list_workspaces_async` method would make this much easier to work with. I'll also look into the pagination work, as I imagine this wouldn't be as useful without